### PR TITLE
Doc - Cluster Architecture Improvements

### PR DESCRIPTION
**Pull request by sleto-it**
**Should be backported to 3.4 after review phase**

This PR aims at improving the "Cluster Architecture" section in particular to:

1. emphasize the need of setting RF manually to have data redundancy, to limit the cases where it is taken for granted that data is redundant when the cluster is in use. This is extremely important for new users
2. "Internally" word is used to emphasize that it is from an internal point of view only that writes and reads are sent to the leader shard, but for the user this is transparent
3. add more details of the failover mechanism - some cases were missing. Now hopefully things are explained in the more general case
4. the Mesos integration is moved to /stash/ . This part needed a change in particular as it included "we use the Apache Mesos infrastructure by default.". Now a section "deployment" is added with very high level info on the fact that the cluster can be deployed in several ways, including dc/os and k8 - and a link to deployment is added for further details 
5. minor typos and style, e.g. use of _DBServer_ instead of DBServer and similar
6. a statement about sync repl (the one that explains that it is still possible to set RF 1) is moved from the section "Shard movement and resynchronization" (where it seems offtopic) to the section "Synchronous replication", to keep info on the same topic aggregated
7. The section "Cluster ID" would benefit from some more details - these are not included in this PR. Would be nice to clarify if such IDs are the ones returned by  GET /_admin/server/id  and are the ones in the format _CRDN-string_, _PRMR-string_ . Leaving this open to comments by reviewers
8. Recommendation about not running Agent instances on the same machines where DBServer instances are running is added
8. Info about the fact that cluster processes are supposed to run within the same DC is included, and DC2DC is mentioned + link added

**Preview:**
- http://web01.arangodb.biz/documentation/cluster-arch-improv/Manual/Architecture/DeploymentModes/Cluster/Architecture.html

### CODE REVIEW

File: Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md [hunk 1]
```diff
@@ -98,6 +100,17 @@ which are suitable for different usage scenarios:
...
+on the same machines where _DBServer_ instances are running.
+
+When deploying using the tool [_Starter_](../../../Deployment/ArangoDBStarter/README.md)
+this can be achived by using the options `--cluster.start-dbserver=false` and
+`--cluster.start-coordinator=false` on the first three machines where the _Starter_
+is started, if the desidered _Agency_ _size_ is 3, or on the first 5 machines
+if the desidered _Agency_ _size_ is 5.
+{% endhint %}
+
 Cluster ID
```
Comment (sleto-it) [inline Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md]: Is this still something current?

Comment (sleto-it) [reply Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md]: asked level2.

**Comment by sleto-it** [comment]
There are two icons in the page, similar to the following that need fix

screenshot 2019-02-06 11 32 24

**Comment by Simran-B** [comment]
@sleto-it The emojis were used deliberately as value for a user-defined attribute. It was not a bug in the documentation generation. I replaced it with `"foo": "bar"`.

**Comment by sleto-it** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-documentation/289/

**Comment by sleto-it** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-documentation/290/

File: Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md [hunk 2]
```diff
@@ -163,86 +215,141 @@ The data will always be stored on the _DBServers_.
...
        "_rev" : "7987"
    }
    ```
-   When a follower fails, the leader will give up on it after 3 seconds
-   and proceed with the operation. As soon as the follower (or the network
-   connection to the leader) is back up, the two will resynchronize and
-   synchronous replication is resumed. This happens all transparently
-   to the client.
+
+Obviously, synchronous replication involves a certain increased latency for
```
Comment (Simran-B) [inline Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md]: How about changing this slightly?
```suggestion
Obviously, synchronous replication comes at the cost of an increased latency for
```

File: Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md [hunk 3]
```diff
@@ -1,29 +1,33 @@
 Cluster Architecture
 ====================
 
-The cluster architecture of ArangoDB is a _CP_ master/master model with no
-single point of failure. With "CP" we mean that in the presence of a
+The Cluster architecture of ArangoDB is a _CP_ master/master model with no
+single point of failure. 
+
+With "CP" we mean that in the presence of a
```
Comment (Simran-B) [inline Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md]: Explain the origin of this two letter word
```suggestion
With "CP" in terms of the [CAP theorem](https://en.wikipedia.org/wiki/CAP_theorem)
we mean that in the presence of a
```

File: Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md [hunk 4]
```diff
@@ -63,14 +67,16 @@ and restarted as needed.
...
-either be leader or follower for a shard.
+_DBservers_ are the ones where the data is actually hosted. They
+host shards of data and using synchronous replication a _DBServer_ may
+either be _leader_ or _follower_ for a shard.
 
 They should not be accessed from the outside but indirectly through the
 _Coordinators_. They may also execute queries in part or as a whole when
 asked by a _Coordinator_.
 
+See [below](#sharding) for more information on sharding.
```
Comment (Simran-B) [inline Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md]:
```suggestion
See [Sharding](#sharding) below for more information.
```

File: Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md [hunk 5]
```diff
@@ -95,9 +101,31 @@ which are suitable for different usage scenarios:
     latency. Essentially, this moves some of the database distribution
     logic to the machine where the client runs.
 
-As you acn see, the _Coordinator_ layer can be scaled and deployed independently
+As you can see, the _Coordinator_ layer can be scaled and deployed independently
 from the _DBServer_ layer.
 
+{% hint 'warning' %}
+It is a best practice and a recommended approach to not run _Agent_ instances
```
Comment (Simran-B) [inline Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md]: One can easily miss the **not** in the sentence, let's write it the other way around!
```suggestion
It is a best practice and a recommended approach to run _Agent_ instances
```

File: Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md [hunk 6]
```diff
@@ -95,9 +101,31 @@ which are suitable for different usage scenarios:
...
+on the same machines where _DBServer_ instances are running.
```
Comment (Simran-B) [inline Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md]:
```suggestion
on different machines than _DBServer_ instances.
```

File: Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md [hunk 7]
```diff
@@ -95,9 +101,31 @@ which are suitable for different usage scenarios:
...
+this can be achieved by using the options `--cluster.start-dbserver=false` and
+`--cluster.start-coordinator=false` on the first three machines where the _Starter_
+is started, if the desired _Agency_ _size_ is 3, or on the first 5 machines
+if the desired _Agency_ _size_ is 5.
+{% endhint %}
+
+{% hint 'info' %}
+The difference instances part of a single Cluster are supposed to be run in the same
+_Data Center_ (DC), with reliable and high-speed network connection between
+all the machines participating to the Cluster.
```
Comment (Simran-B) [inline Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md]: I don't understand what you try to say here, and it looks like there are some grammar issues as well?

File: Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md [hunk 8]
```diff
@@ -95,9 +101,31 @@ which are suitable for different usage scenarios:
...
+
+Multi-datacenter Clusters, where the entire structure and content of a Cluster located
+in a specific DC is replicated to others Clusters located in different DCs, are
+possible as well. See [Datacenter to datacenter replication](../DC2DC/README.md)
+(DC2DC) for further details.
```
Comment (Simran-B) [inline Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md]: Is this really officially supported? It sounds like you could use DC2DC to create a distributed cluster, which it is not. Also, do we really support 1:N datacenters in DC2DC? In a discussion with Omar our understanding was that you can use it for 1:1 async replication between two clusters, in a single direction (one way).

File: Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md [hunk 9]
```diff
@@ -144,17 +172,41 @@ is synchronous.
 Synchronous replication works on a per-shard basis. Using the option _replicationFactor_,
 one configures for each _collection_ how many copies of each _shard_ are kept in the Cluster.
 
+{% hint 'danger' %}
+If a collection have a _replication factor_ equal to 1, its data is **not**
```
Comment (Simran-B) [inline Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md]:
```suggestion
If a collection has a _replication factor_ of 1, its data is **not**
```

File: Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md [hunk 10]
```diff
@@ -144,17 +172,41 @@ is synchronous.
...
+replicated in other _DBServers_. This exposes you to a risk of data loss, if
```
Comment (Simran-B) [inline Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md]:
```suggestion
replicated to other _DBServers_. This exposes you to a risk of data loss, if
```

File: Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md [hunk 11]
```diff
@@ -144,17 +172,41 @@ is synchronous.
...
+the machine where the _DBServer_ that is storing this collection permanently
```
Comment (Simran-B) [inline Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md]:
```suggestion
the machine running the _DBServer_ with the only copy of the data fails permanently.
```

File: Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md [hunk 12]
```diff
@@ -144,17 +172,41 @@ is synchronous.
...
+fails.
```
Comment (Simran-B) [inline Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md]:
```suggestion
```

File: Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md [hunk 13]
```diff
@@ -144,17 +172,41 @@ is synchronous.
...
+{% endhint %}
+
+{% hint 'warning' %} 
+The _replication factor_ has to be set to a value equals or higher than 2
+to achieve minimal data redundancy via the synchronous replication.
```
Comment (Simran-B) [inline Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md]: Shouldn't this be part of the previous hint?

File: Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md [hunk 14]
```diff
@@ -144,17 +172,41 @@ is synchronous.
...
+{% endhint %}
+
+{% hint 'warning' %}
+An equal-or-higher-than 2 _replication factor_ has to be set **explicitly**
+when the collection is created, or can be set later at run time if you forgot
+to set it at creation time.
```
Comment (Simran-B) [inline Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md]: This should be part of the previous hint at least. Or what is the value in having the information in different hints?

File: Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md [hunk 15]
```diff
@@ -144,17 +172,41 @@ is synchronous.
...
+{% endhint %}
+
+{% hint 'tip' %}
+When using a Cluster, please make sure all the collections that are important
+(and should not be lost in any case) have a _replication factor_ equal or higher
+than 2.
```
Comment (Simran-B) [inline Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md]: This hint seems pretty redundant

File: Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md [hunk 16]
```diff
@@ -286,44 +385,15 @@ being able to use a single technology for various roles in a project.
...
+Deployment
+----------
+
+An ArangoDB Cluster can be deployed in several ways, e.g. by manually
+starting all the needed instances, by using the tool 
+[_Starter_](../../../Programs/Starter/README.md), in
+Docker, in Mesos or DC/OS, and in Kubernetes.
+
+See the [Cluster Deployment](../../../Deployment/Cluster/README.md)
+for instructions.
```
Comment (Simran-B) [inline Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md]:
```suggestion
chapter for instructions.
```

File: Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md (hunk 7 above)
Comment (sleto-it) [reply Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md]: discussed on phone - commited a possible change.  Is it better now, can we consider this resolved? thanks

File: Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md (hunk 14 above)
Comment (sleto-it) [reply Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md]: done, thanks

File: Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md (hunk 3 above)
Comment (sleto-it) [reply Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md]: thanks simran

File: Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md (hunk 13 above)
Comment (sleto-it) [reply Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md]: done - the previous one is a danger one though.

File: Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md (hunk 15 above)
Comment (sleto-it) [reply Documentation/Books/Manual/Architecture/DeploymentModes/Cluster/Architecture.md]: Indeed - trying to communicate something important. I put everything into a single danger hint - let's see how it looks.

**Comment by sleto-it** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-documentation/300/

**Comment by sleto-it** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-documentation/301/

**Comment by sleto-it** [comment]
This is now ready for technical review by Cluster Team Leadership. We have done our best to improve this doc part and clarify crucial points. I will **not** merge without their additional review. Thanks,

File: Documentation/Books/stash/ClusterArchitecture.md [hunk 17]
```diff
@@ -0,0 +1,39 @@
+
+Apache Mesos integration
+------------------------
+
+For the distributed setup, we use the Apache Mesos infrastructure by default.
```
Comment (mchacki) [inline Documentation/Books/stash/ClusterArchitecture.md]: I think this is not true any more?

Comment (sleto-it) [reply Documentation/Books/stash/ClusterArchitecture.md]: (resolved as part of stash folder)

**Pull request merged**